10up/ElasticPress
# Handle expired or missing tokens for orders autosuggest

### Description of the Change

Adds handling for missing or expired tokens. Now if an search fails a new token will be fetched and the search will be reattempted. This will handle situations where a token has expired. Refreshing a token will only be attempted once.

Additionally I've included a few fixes for the suggestion dropdown:

- Fixed an issue where the date and time of orders could be incorrect.
- Added support for opening suggestions in a new tab by holding Control/Command when selecting a suggestion.
- Close the suggestions dropdown immediately when the input is empty, instead of debouncing.


Closes #3296 

### How to test the Change
From #3175:
1. Connect to an EP.io environment that supports the order search API.
2. Activate the WooCommerce and Protected Content features.
3. Add `( new Orders() )->setup();` to the `setup()` method of the WooCommerce feature class.
4. Perform an index.
5. Visit _WooCommerce > Orders_ and start searching for orders. Suggestions should be offered in a listbox below the search form.

New:
- Watching the network tab, visit _WooCommerce > Orders_. A request to `wp-json/elasticpress/v1/token` should be sent and should return a token.
- Start searching for orders. Suggestions should be offered in a listbox below the search form.
- In the database, in `wp_usermeta`, find `ep_token` for your user and manipulate it so that it's invalid.
- Watching the network tab, visit _WooCommerce > Orders_. A request to `wp-json/elasticpress/v1/token` should be sent and should return the manipulated token.
- Watching the network tab, start searching for orders. The first request to the ElasticPress.io API should fail. Another request to `wp-json/elasticpress/v1/token` should be sent and should return a new token. Another API request to ElasticPress.io should follow and should succeed. Suggestions should now be offered in a listbox below the search form.

### Changelog Entry
See #3175

### Credits
Props @JakePT 

### Checklist:
